flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/internal/converter/AbstractJdbcRowConverter.java

```diff
@@ -18,54 +18,255 @@
+				array[i] = elementConverter.deserialize(objects[i]);
+			}
+			return new GenericArrayData(array);
+		};
 	}
 
 	/**
-	 * Create a runtime JDBC field converter from given {@link LogicalType}.
+	 * Create a nullable JDBC f{@link JdbcSerializationConverter} from given sql type.
 	 */
-	public JdbcFieldConverter createConverter(LogicalType type) {
-		LogicalTypeRoot root = type.getTypeRoot();
-
-		if (root == LogicalTypeRoot.SMALLINT) {
-			// Converter for small type that casts value to int and then return short value, since
-	        // JDBC 1.0 use int type for small values.
-			return v -> ((Integer) v).shortValue();
-		} else {
-			return v -> v;
+	protected JdbcSerializationConverter createNullableExternalConverter(LogicalType type) {
+		return wrapIntoNullableExternalConverter(createExternalConverter(type), type);
+	}
+
+	protected JdbcSerializationConverter wrapIntoNullableExternalConverter(JdbcSerializationConverter jdbcSerializationConverter, LogicalType type) {
+		final int sqlType = JdbcTypeUtil.typeInformationToSqlType(TypeConversions.fromDataTypeToLegacyInfo(
+			TypeConversions.fromLogicalToDataType(type)));
+		return (val, index, statement)  -> {
+			if (val == null || val.isNullAt(index) || LogicalTypeRoot.NULL.equals(type.getTypeRoot())) {
+				statement.setNull(index + 1, sqlType);
+			} else {
+				jdbcSerializationConverter.serialize(val, index, statement);
+			}
+		};
+	}
+
+	protected JdbcSerializationConverter createExternalConverter(LogicalType type) {
+		switch (type.getTypeRoot()) {
+			case BOOLEAN:
+				return (val, index, statement) -> statement.setBoolean(index + 1, val.getBoolean(index));
+			case TINYINT:
+				return (val, index, statement) -> statement.setByte(index + 1, val.getByte(index));
+			case SMALLINT:
+				return (val, index, statement) -> statement.setShort(index + 1, val.getShort(index));
+			case INTEGER:
+			case INTERVAL_YEAR_MONTH:
+				return (val, index, statement) -> statement.setInt(index + 1, val.getInt(index));
+			case BIGINT:
+			case INTERVAL_DAY_TIME:
+				return (val, index, statement) -> statement.setLong(index + 1, val.getLong(index));
+			case FLOAT:
+				return (val, index, statement) -> statement.setFloat(index + 1, val.getFloat(index));
+			case DOUBLE:
+				return (val, index, statement) -> statement.setDouble(index + 1, val.getDouble(index));
+			case CHAR:
+			case VARCHAR:
+				// value is BinaryString
+				return (val, index, statement) -> statement.setString(index + 1, val.getString(index).toString());
+			case BINARY:
+			case VARBINARY:
+				return (val, index, statement) -> statement.setBytes(index + 1, val.getBinary(index));
+			case DATE:
+				return (val, index, statement) ->
+					statement.setDate(index + 1, Date.valueOf(LocalDate.ofEpochDay(val.getInt(index))));
+			case TIME_WITHOUT_TIME_ZONE:
+				return (val, index, statement) ->
+					statement.setTime(index + 1, Time.valueOf(LocalTime.ofSecondOfDay(val.getInt(index) / 1000L)));
```

Comment:
Would be better to use `LocalTime.ofNanoOfDay(val.getInt(index) * 1_000_000L)` to support millisecond.